### ISSUE TITLE
extend src/bindings/jna/hello/src/main/java/HelloElektra.java

### ISSUE THREAD

**Issue by markus2330** [ElektraInitiative/libelektra#4093]
src/bindings/jna/hello/src/main/java/HelloElektra.java should include examples from all parts of the API

**Comment by leothetryhard** [comment]
We (@lukashartl and @leothetryhard) want to work on this issue. 
 I compiled elektra with `-DBINDINGS="ALL;-DEPRECATED"` which to my understanding includes `jna`.
I tried running the `HelloElektra.java` by following the [JNA docs](https://www.libelektra.org/bindings/jna) and executing `./gradlew hello:run` in `/src/bindings/jna`. However the build fails with the following error: 

![jna-error](https://user-images.githubusercontent.com/25547927/160278260-e59a7d3a-8d96-41e8-847a-23222bd7cca2.png)

Can you point me towards what I'm missing?

**Comment by kodebach** [comment]
Could you please try again... The error basically means that Gradle couldn't find the dependency `com.thoughtworks.xstream` (which is required by the `gradle-versions` plugin, we use). As far as I can tell, the dependency is available in the standard repository and for me `./gradlew hello:run` just worked. Maybe it was a weird HTTP timeout or something similar...

If the error persists, I'll look into it again

**Comment by leothetryhard** [comment]
Yes it works now. `pluings.gradle.org` must have been down when I first tried. Thank you!

**Comment by leothetryhard** [comment]
[CM T3] Update: There are new issues when trying to run HelloJava. 
When executing `./gradlew hello:run` the build fails in example 5:
![hellojava_error_build](https://user-images.githubusercontent.com/25547927/170306387-cb6797bd-d3e4-40c9-adff-127b6f2fa858.png)

The line of concern is `Key k = ks.lookup(key).orElseThrow(AssertionError::new);`
My colleague @lukashartl was able to reproduce the problem.
Has there been recent changes to `.lookup()`?
Or maybe is it something trivial we are overlooking?
Help would be appreciated @markus2330 @kodebach in order to continue with the work.

**Comment by leothetryhard** [comment]
#4378 implemented:

 * [x]  Check whether the keys are already available in the key database
 * If not, save them to the key database
 * If they are available, load their values

 * [x]  Update meta data on a key
 * [x]  Use Key iterator to dump meta data

**Comment by kodebach** [comment]
I'm not 100% sure without looking at the code, whether or not the current `kdbGet` would clear the key. In the `new-backend` branch IIRC I took the approach that after `kdbGet` the part below `parentKey` matches exactly what is persisted in the KDB. Otherwise, you'd need to manually clear the `KeySet` before calling `kdbGet` to make sure you actually get the persisted config, if e.g. a key was removed by another process between two `kdbGet` calls.

> Of course this does not mean, it is a perfect example ;)

Yes, even if it works. It is very confusing as a first example. Especially without a comment stating e.g. "this works because ...".

**Comment by tucek** [comment]
I think the best course of action is to improve the example by actually writing a key to KDB and then fetching it and removing it afterwards. do you agree?

**Comment by markus2330** [comment]
This would also tests many other things which you might actually to avoid to test :wink:

**Comment by tucek** [comment]
Could we then just try to read a key that is guaranteed to be already there? Would you have any candidates in mind?

**Comment by kodebach** [comment]
The docs say `kdbGet` only appends, but there could very well be a bug. There are definitely some `ksClear` calls in `kdbGet`, so it might be that the `KeySet` is cleared before the keys are appended. 


https://github.com/ElektraInitiative/libelektra/blob/a5b8684ed10a939b94597e587730ae5b26b09128/src/libs/elektra/kdb.c#L1426

https://github.com/ElektraInitiative/libelektra/blob/a5b8684ed10a939b94597e587730ae5b26b09128/src/libs/elektra/kdb.c#L1476

**Comment by markus2330** [comment]
> Could we then just try to read a key that is guaranteed to be already there? Would you have any candidates in mind?

system:/elektra/version/infos/licence contains BSD

> There are definitely some ksClear calls in kdbGet, so it might be that the KeySet is cleared before the keys are appended.

Is this still the problem in the new-backend branch? We should clean it up, can you please create an issue? (@kodebach)

**Comment by kodebach** [comment]
> Is this still the problem in the new-backend branch?

To quote myself from [above](https://github.com/ElektraInitiative/libelektra/issues/4093#issuecomment-1162019694):

> In the `new-backend` branch IIRC I took the approach that after `kdbGet` the part below `parentKey` matches exactly what is persisted in the KDB. Otherwise, you'd need to manually clear the `KeySet` before calling `kdbGet` to make sure you actually get the persisted config, if e.g. a key was removed by another process between two `kdbGet` calls.

It's also described this way in the updated documentation in `new-backend` (some docs may be outdated).

**Comment by markus2330** [comment]
Yes, exactly, there was a "IIRC" in the sentence and there might be outdated docs.

I agree that kdbGet should only append, so `ksClear (ks)` in the code is fishy. If the problem might be in the new-backend branch, we need to investigate (create a new issue etc.). If it is gone in new-backend, we do not need to worry as time (once new-backend gets merged) will fix the issue.

To come to 1.0 we need to be more careful about such issues. Hints about important bugs must be taken seriously. And that the whole config of the user gets cleared (which might be persisted in a later kdbSet) is a very serious bug.

**Comment by kodebach** [comment]
> that the whole config of the user gets cleared (which might be persisted in a later kdbSet) is a very serious bug

I don't think that this is what happens. Something like that would have been spotted pretty quickly. If `ksClear` is actually used to clear parts of `ks` (and not some temporary copy), then I'm sure these parts will then be replaced by the current contents of the KDB (that's what happens in `new-backend`).

In the `new-backend` branch we use `ksClear` exactly, because there may be a later `kdbSet`. Take this scenario for example:

1. Application `A` calls `kdbGet` with `ks`
2. Application `B` deletes a key from the KDB
3. Application `A` calls `kdbGet` again with the same `ks` (to get the new state of the KDB)
4. Application `A` does some modifications to `ks` and calls `kdbSet` to persist them

If `kdbGet` _only appends_, then the key that `B` deleted will be inserted again by `kdbSet` in 4. That's why in `new-backend` anything that would overlap with a mountpoint loaded by `kdbGet` is removed from `ks`. All other keys in `ks` remain, since they won't be persisted by `kdbSet`.

In the _only appends_ version, a caller would have to pass an empty `ks` (i.e. call `ksClear(ks)` first) to ensure the state of the KDB is correctly reflected after `kdbGet`.

Note: Adding default fallbacks before calling `kdbGet` _should_ still work in `new-backend`. You can add `default:/` keys. These will never overlap with a mountpoint (since adding `default:/` mountpoints is forbidden) and therefore won't be cleared (they also won't be persisted, but that's a good thing). If this doesn't work in `new-backend` right now, then that's a bug.

> I agree that kdbGet should only append

I'm not sure that would be the best solution, but if you think it is better than the current approach in `new-backend`. Then we can change it.
